File: 0238-product-of-array-except-self/solution.cpp
class Solution {
public:
    vector<int> productExceptSelf(vector<int>& nums) {
        
        int ctz = 0 ; 
        int pro = 1 ;

        for(auto &i : nums){
            if(i != 0){
             pro*= i;
            }else ctz++ ;
        }

        vector<int> res(nums.size() , 0) ;
        if(ctz > 1) return res ;
        else if( ctz == 1){

            for(int i = 0 ; i < nums.size() ; i++){
                if(nums[i] != 0 ){
                    res[i] = 0 ;
                }
                else res[i] = pro;
            }

        }
        else {
            for(int i = 0 ; i < nums.size() ; i++){
                res[i] = pro/nums[i];
            }

        }
        return res ;

    }
};

File: 0238-product-of-array-except-self/solution.java
class Solution {
    public static int[] productExceptSelf(int[] nums) {
        int n = nums.length;
        int countZero = 0, mulValue = 1, lastZeroIndex = -1;
        int[] pre = new int[n+1]; pre[0] = 1 ;
        int[] suf = new int[n+1]; suf[n] = 1 ;

        for (int i = 0; i < nums.length; i++) {
            pre[i+1]= pre[i]*nums[i];
            suf[n-1-i] = suf[n-i]*nums[n-i-1];
            if (nums[i] == 0) {
                countZero++;
                lastZeroIndex = i;
            } else {
                mulValue *= nums[i];
            }
        }

        if (countZero > 1) {
            Arrays.fill(nums, 0);
            return nums;
        } else if (countZero == 1) {
            Arrays.fill(nums, 0);
            nums[lastZeroIndex] = mulValue;
        } else {
            for (int i = 0; i < n; i++) {
                nums[i] = pre[i]*suf[i+1];
            }
        }
        return nums;
    }
}
